lib/pharos/kube/resource.rb

```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+require 'kubeclient'
+require 'pathname'
+
+module Pharos
+  module Kube
+    class Resource
+      def initialize(session, resource)
+        @resource =
+          case resource
+          when Kubeclient::Resource
+            resource
+          when Hash
+            Kubeclient::Resource.new(resource)
+          else
+            raise TypeError, "Expected an instance of KubeClient::Resource or Hash"
+          end
+        @api_version = @resource.apiVersion || 'v1'
+        @client = session.resource_client(@api_version)
+        freeze
+      end
+
+      # @return [Kubeclient::Resource]
+      def apply
+        update
+      rescue Kubeclient::ResourceNotFoundError
+        create
+      end
+
+      # @return [Kubeclient::Resource]
+      def create
+        @client.create_resource(@resource)
+      end
+
+      # @return [Kubeclient::Resource]
+      def update
+        @client.update_resource(@resource)
+      end
+
+      # @return [Kubeclient::Resource]
+      def get
+        @client.get_resource(@resource)
+      end
+
+      # @return [Kubeclient::Resource,FalseClass]
+      def delete
+        if metadata.selfLink
+          api_group = metadata.selfLink.split("/")[1]
+          path = metadata.selfLink.gsub("/#{api_group}/#{@api_version}", '')
+          @client.rest_client[path].delete
+        else
+          definition = entity_definition
+          @client.get_entity(definition.resource_name, metadata.name, metadata.namespace)
+          @client.delete_entity(
+            definition.resource_name, metadata.name, metadata.namespace,
+            kind: 'DeleteOptions',
+            apiVersion: 'v1',
+            propagationPolicy: 'Foreground'
+          )
+        end
+      rescue Kubeclient::ResourceNotFoundError
+        false
+      end
+
+      def metadata
```

Comment:
What about accessing the actual `spec` or other data on the `@resource`?

It seems that `Pharos::Kube::Resource` is currently only usable for writing resources to kube, but not reading them back from kube. This resource wrapper currently doesn't work for e.g. something like `ConfigureNetwork#ensure_passwd` or the `Pharos::Kube::CertManager` needs to get at the resource secrets and status... 

Not entirely sure what the fix should be, though, because this mixes up the verb actions like `get` or `delete` with the attributes like `data`, `spec` or `status`... maybe those could be accessed via `def []` or something?

Reply:
Also, I don't think we can just write out accessors for all the different resource type fields either... different resource types will have different fields in addition to the generic `kind`, `apiVersion` and `metadata`... and custom resources types can have any schema at all.

Reply:
The `.get` could wrap the responses inside a `Pharos::Kube::Resource`.

Then you could do `x = @session.resource(foofoo).get; x.delete if x.foofoo.nil?`

It could have `[]` and `method_missing` which would still allow access to `resource[:get]` in case you need `resource.get`.

Some orms have `.attributes`, that could work too.